### ISSUE TITLE
Completed task gets re-indexed if edited.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#23883]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create a task
2. Set to completed
3. Open task report and write something & watch the sidebar
## Expected Result: 
Completed task should not be reindex and should not move to top to the report list or if it has to move up then it should persist its index.
## Actual Result: 
Completed tasks moves to top while writing something in composer and gets back to its position if the message is sent.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.47-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/25981bb9-381d-4922-bec4-94e05d450016

https://github.com/Expensify/App/assets/43996225/7a788179-250b-4cff-a9b7-218d4756b04e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @krishna2323
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690436338235719

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by niteshletxsoft** [comment]
### **Proposal**

### **Please re-state the problem that we are trying to solve in this issue.**
Completed task should not be reindex and should not move to top to the report list or if it has to move up then it should persist its index.

### **What is the root cause of that problem?**
The root cause of this issue is the ordering of the report ID's based on their status and when the completed task has a draft comment, According to the current logic all draft report comes on the top.

 https://github.com/Expensify/App/blob/main/src/libs/SidebarUtils.js#L142-L145

This is the part of the code, which is causing the current issue.

### **What changes do you think we should make in order to solve the problem?**
The actual solution to this issue is to keep the Completed task at their position even if the task has a draft comment.

We can check if this REPORT is a task and completed one, We will not put the REPORT in the draftReports array.

```javascript
// The completed tasks should not move the position when have a draft
if (report.hasDraft && !(ReportUtils.isTaskReport(report) && ReportUtils.isCompletedTaskReport(report))) {
    draftReports.push(report);
    return;
}
```

As per the another half comment of the issue by the reporter "if it has to move up then it should persist its index."
We can not update the completed task position on new comments because this will conflict with the logic of archivedReports, which keeps the completed task at the bottom of the list.

**Comment by twisterdotcom** [comment]
I agree, this is intentional. Not a bug.

**Comment by thienlnam** [comment]
I had the same initial view but in the video, it looks like after you send the message it is reverting it’s position which is the bug here. I agree it reindexing while in draft state is correct, but it seems like after the message is sent and it is no longer in draft state it should stay at the top

**Comment by twisterdotcom** [comment]
Hmmm, yes I haven't not used focus mode for a long time. I'll retest this.

**Comment by niteshletxsoft** [comment]
Yes, You are right @thienlnam. either the completed task should change the position on draft and keep the position after sending the message or should not change the position when has the draft message.
As shown in the bug, It changes the position when has the draft message and come back to original position when message is sent.

According to me, It should not come to top position with draft message as the task has been completed and that's the solution i have provided.

**Comment by twisterdotcom** [comment]
>I think we should not consider a completed task as a Archived report but Non Archived Report instead

This makes sense to me.

**Comment by twisterdotcom** [comment]
@parasharrajat and @thienlnam - can we assign @dukenv0307 on this one?

**Comment by thienlnam** [comment]
Proposal looks good - initially I thought we would close the task from being able to chat on and it would be "done". However, things have changed since and now the task report should always stay open

**Comment by niteshletxsoft** [comment]
@thienlnam 
What do you think about my proposal. Simple way to handle this issue.
When the user type in a completed task, The chat does not change it's position and stay where it was.

https://github.com/Expensify/App/issues/23883#issuecomment-1657249588

**Comment by thienlnam** [comment]
I prefer the recommended proposal as it addresses the root cause - anyways, leaving up to the C+ and internal engineer assigned on this issue

**Comment by twisterdotcom** [comment]
@srikarparsi just waiting on you.

**Comment by srikarparsi** [comment]
Sorry for the delay, was OOO friday. @dukenv0307's proposal looks good to me as well.

**Comment by twisterdotcom** [comment]
>I noticed that all reports change position when drafts are added/removed from them

This is just because the draft moves _up_ in the hierarchy though right? When the message is sent, it doesn't move back down again. If so, I think this is fine.

**Comment by twisterdotcom** [comment]
You shouldn't have to send a proposal on Upwork anymore. We should automatically send you a contract if we have your Upwork profile saved. Only proposals in GitHub.

**Comment by thienlnam** [comment]
@Krishna2323 Can you please email the email address you linked to contributors@expensify.com and we can take a look from there?

**Comment by thienlnam** [comment]
@Krishna2323 I took a look, and our automation doesn't work with the Upwork custom profile links. You'll have to update your stored upwork link to your regular profile link. Can you paste the below in a comment?

```
Contributor details
Your Expensify account email: belivethatkg@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~011653c0d62a6dc458
```

**Comment by twisterdotcom** [comment]
Payment Summary:
- @parasharrajat will be paid $1500 via newDot
- @dukenv0307 has been paid $1500 via Upwork: https://www.upwork.com/nx/wm/workroom/34468017/overview
- @Krishna2323 will be paid $250 here: https://www.upwork.com/nx/wm/offer/26179979 (just accept the offer! but also let's store your Upwork details for future too following @thienlnam's instructions).

**Comment by JmillsExpensify** [comment]
Reviewed the details for @parasharrajat. Approved for $1,500 payment in NewDot based on the BZ summary above.
